runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -222,6 +227,35 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
     return Optional.<Boolean>absent();
   }
 
+  boolean checkMaxWatermark(DataflowPipelineJob job) throws IOException {
+    JobMetrics metrics = options.getDataflowClient().projects().jobs()
```

Comment:
what about exceptions?
retries will reduce the flakiness, when the backend is temporarily unavailable.